### ISSUE TITLE
Extra Libraries

### ISSUE THREAD

**Issue by cianyleow** [HPCE/hpce-2015-cw6#5]
Are we allowed to install/use any extra libraries? If so, how do we go about doing this?

**Comment by m8pple** [comment]
I think it depends on the specific purpose of the library. Anything that is providing
generic maths functions or compute services I'm inclined to allow. If it is targeting
the problem very specifically (e.g. a fractal rendering library), then I think it
is not in the spirit of the task.

For generic libraries it needs to be part of the shared image - I allowed for
per-submission provisioning scripts last year, and it was a nightmare on my
end, which is why I haven't mentioned it this year.

So if you have a specific `sudo apt-get install ...` command that will work
in the AMI (and it isn't `libcudafractal`) then I can add it to the single setup
script I have for marking.

For now, let's say that you just post the `apt-get` command (or a
`wget/configure/make/make install`) to this issue - so if you want to use a
library, I'm afraid you have to leak that information to others.

I'll leave the issue open.

**Comment by cianyleow** [comment]
Sorry, forgot to actually reply to this.

Could we add the libjpeg-turbo library? It is essentially libjpeg with SIMD instructions and some functionality to make it a little faster. 

Jake and I compiled and built it for DoC machines (as we don't have sudo rights), so I can either make a script for that or I think the command should be `sudo apt-get install libjpeg-turbo8` instead of `sudo apt-get install libjpeg8`. Let me double check that on an AWS instance (Jake and I are being careful about what we start up now...) and I'll let you know if we need anything different if the above is allowed.

One caveat is that I think you need to uninstall libjpeg, but it does provide all of the functionality of libjpeg, but with the speed ups - so possibly a bonus for everyone in the course? If you don't want to swap out libjpeg for libjpeg-turbo, we can just include the shared object files needed and ensure that it compiles and runs with out affecting the normal libjpeg (have to change the LD_LIBRARY_PATH for dynamic linking at one point but only for our session and can be done in the makefile).

**Comment by m8pple** [comment]
I'm aware of libjpeg-turbo, and deliberately didn't install it. It's not that I'm against you
using it, but it's not something I want people to be able to globally take advantage
of. It takes some thought and analysis to determine the bottlenecks, so only those
who do it should get the benefit.

So I think here a local install would make sense as part of your makefile or environment,
as it isn't something that needs system level permissions.

I will commit to having your `bin` directory listed in LD_LIBRARY_PATH whenever I
run your executable, and I will not move your `bin/julia_filter` after the build. So during
`make bin/julia_filter` you can do whatever setup is needed, and can also handle
the link command needed to pick up the different jpeg lib.

I'll be doing a couple of test runs today and tomorrow (I'll email out a schedule), so
there will be a chance to fine-tune it.

**Comment by cianyleow** [comment]
Thanks, will close this issue now then.

**Comment by cianyleow** [comment]
Hi, found that to compile libjpeg-turbo, some more tools are required. Could you run `sudo apt-get -y install curl autoconf automake libtool nasm yasm` before making ours?

**Comment by m8pple** [comment]
Yes, I've added those libraries to the provision script. I'll document them on the front-page too.
